Completed regtest and signet tests to pass and run automatically and added script to run tests with one command

solarsailorneo opened the pull request:
+ Config switch for regnet and signet
+ Added script for running automatic testing
+ Need to test signet with credits
+ Modified README to be more explicit

solarsailorneo committed:
Completed regnet tests and added script to run tests automatically

+ Config switch for regnet and signet
+ Added script for running automatic testing
+ Need to test signet with credits
+ Modified README to be more explicit

solarsailorneo commented:
### Summary of Improvements

Greetings, kasarlabs maintainers. I'd like to provide a breakdown of the enhancements and alterations I introduced to the `bitcoin-da-rs` crate:

1. **Handling of the "bark" Identifier**: 
   - I identified that the identifier "bark" wasn't being addressed in some parts of the codebase. It wasn't immediately clear whether this omission was a design decision or a latent bug. I've made adjustments to ensure that the "bark" identifier is now uniformly processed across the relevant functions. I'd appreciate it if the team could confirm if this change aligns with the intended design or if it was indeed an oversight that needed rectification.

2. **Addition of a Block Generation Function for Regnet**:
   - To facilitate testing and streamline operations on the regnet network, I added a function dedicated to generating blocks on regnet. This enhancement should aid developers in swiftly setting up a testing environment without manually having to mine blocks.

3. **Refactoring**:
   - I made a review of several test functions within the library and identified potential areas for more clarity. Some of the function logic has been refactored for improved clarity. I think these modifications should lead to a modest boost in the crate's performance and maintainability.

4. **Integration**:
   - I have design an automatic test paradigm to be able to run tests in regnet with one script. The version for signet is not totally finished yet.

I hope these details offer clarity regarding the improvements introduced. Please give me your feedback and suggestions. If any aspect needs further elaboration or adjustment, let me know.

0xEniotna commented on run_tests.sh:
```diff
@@ -0,0 +1,140 @@
+done
+
+# Copy the bitcoin configuration from the current directory to ~/.bitcoin
+cp ./bitcoin.conf ~/.bitcoin/
```
Not sure that's usefull, some people might already have a bitcoin.conf. 
The file is an example.

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -323,61 +341,96 @@ mod tests {
 
+
     #[test]
     fn test_reveal2() {
```
unnecessaries changes here, apart from the NODE_IP

solarsailorneo replied:
Will roll back except for NODE_IP and given, when, then comments

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -323,61 +341,96 @@ mod tests {
             .unwrap();
-        // assert_eq!(network, Network::Regtest);
-        assert_eq!(network, Network::Signet);
+        assert_eq!(network, Network::Regtest);
```
```suggestion
        // Change this, if required
        // assert_eq!(network, Network::Signet);
        assert_eq!(network, Network::Regtest);
```

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -323,61 +341,96 @@ mod tests {
     #[test]
     fn test_write() {
+        // ===============================
+        // Given: a configured Bitcoin relayer on the REGNET network
```
Would be better to keep both ways signet/regtest. 
At least for us, it is important to be able to switch fast aka commenting/uncommenting.
Ugly but efficient

solarsailorneo replied:
Made changes and was able to run `./run_tests.sh -t test_write` (also ran `cargo test test_write`) succesfully for both regtest and signet. Do you still need the ability to comment in an out after this?

0xEniotna replied:
Nop, looks good like that !

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -323,61 +341,96 @@ mod tests {
+        // When: data is written to the network
+        // ==================================
+
+        let write_result = relayer.write(embedded_data);
```
unnecessary matching pattern change here but it's in a test, we can keep it :)

solarsailorneo replied:
sounds good. Yeah, I always go back and forth in these situations with tests lol

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -386,47 +439,149 @@ mod tests {
-                // Change this line to whatever data you want.
-                // "bark" is appended to the beginning of the data
-                assert_eq!(data, b"barkHello, world!".to_vec());
+                assert!(data.windows(b"barkHello, height!".len()).any(|window| window == b"barkHello, height!"));
```
not sure to understand why

solarsailorneo replied:
When I use:
`assert_eq!(data, b"barkHello, world!".to_vec());`

I get:
> thread 'tests::test_read_data_by_block_height' panicked at 'assertion failed: `(left == right)`
>   left: `[98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 104, 101, 105, 103, 104, 116, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33, 98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33]`,
>  right: `[98, 97, 114, 107, 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33]`', crates/bitcoin-da/tests/bitcoin_da_tests.rs:490:17

It seems like there is more data than just "barkHello, world!". Not sure if I'm doing something incorrectly on my end.

0xEniotna replied:
looks like it's the same data in loop, i'm not sure why

solarsailorneo replied:
Changed back to what you guys had in your `test_read_height` function:

>         // Change this line to whatever data you want.
>                 // "bark" is appended to the beginning of the data
>                 assert!(
>                     data == b"barkHello, world!".to_vec(),
>                     "Assertion failed. This test is designed to be run manually. Expect it to fail in auto mode. Received data: {:?}",
>                     data
>                 );
>                 println!("Successful read");

It works well when I run the test individually, but when I run all the tests seems like things are getting written to the block I'm trying to use from other tests. Will look into it

solarsailorneo replied:
I think I understand what is going on (Correct me if I'm wrong). The `read_height` function returns an appended vector of all transaction data from a single block. But the data we are sending could just be one of many transaction data, as opposed to the only piece of data in there. 

Assuming this is true, I decided to use a unique identifier in the embedded data to make sure we are retrieving only the data we sent.

As of now, changed back to this:
```
            Ok(data) => {
                // Check for the presence of the unique data in the block
                assert!(
                    data.windows(embedded_data.len()).any(|window| window == embedded_data.as_slice()),
                    "Unique data not found in block. Received data: {:?}",
                    data
                );
                println!("Successful read");
            }
            Err(e) => panic!("Read failed with error: {:?}", e),
```

0xEniotna commented on crates/bitcoin-da/src/lib.rs:
```diff
@@ -1,5 +1,6 @@
 use bitcoin::amount::Amount;
 use bitcoin::blockdata::script::Builder;
+// use bitcoin::blockdata::script::Instruction;
```
remove comment

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -386,47 +439,149 @@ mod tests {
+
     #[test]
-    fn test_read_height() {
+    fn test_read_data_by_block_height() {
```
great change for regtest !
Will not work for signet unfortunately, that's why we preferred to hardcode the block height.

solarsailorneo replied:
I can have a single test where we can have both the check for regtest and the hardcode for blockheight and keep the easy toggle at the command line level. Would this be something that would make sense for you guys to have?

0xEniotna replied:
I'm good with that !

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -386,47 +439,149 @@ mod tests {
+    
+    #[cfg(feature = "regnet")]
+    #[test]
+    fn test_read_transaction_from_mempool() {
```
Here, you added two tests to read the data from a txid in the mempool. 
The test we had was to read the data from a txid AND a block height, we should still have it somewhere.

Good tests tho

solarsailorneo replied:
Placed them back in

solarsailorneo replied:
Placed them back in

0xEniotna commented:
Good PR ! Some remarks tho:
in the run_test.sh file, I'd prefer not to do anything that touches the nodes. I'd rather have the user configure it by himself. One could already have a bitcoin node running. We should assume that the user already has a bitcoind running somewhere.

I also wrote some comments in the code

solarsailorneo committed:
Update crates/bitcoin-da/tests/bitcoin_da_tests.rs

Co-authored-by: Antoine <101047205+0xEniotna@users.noreply.github.com>

solarsailorneo commented:
> Good PR ! Some remarks tho: in the run_test.sh file, I'd prefer not to do anything that touches the nodes. I'd rather have the user configure it by himself. One could already have a bitcoin node running. We should assume that the user already has a bitcoind running somewhere.
> 
> I also wrote some comments to address in the code

Thank you.

When you say: "I'd rather have the user configure it by himself.", does this include things like:
`bitcoin-cli createwallet test`, `bitcoin-cli loadwallet test`, etc?

0xEniotna commented on run_tests.sh:
```diff
@@ -0,0 +1,120 @@
+done
+
+# Derive the NETWORK value from ~/.bitcoin/bitcoin.conf
+if grep -q '^regtest=1' ~/.bitcoin/bitcoin.conf; then
```
smart !

0xEniotna commented on run_tests.sh:
```diff
@@ -0,0 +1,120 @@
+      TEST_NAME="$OPTARG"
+      ;;
+    L)
+      LONG_TESTS=1
```
what is long test ?

solarsailorneo replied:
Long tests are the tests I added. When running in signet they take a long time to run, as I am building the transaction and then checking it.

0xEniotna commented on run_tests.sh:
```diff
@@ -0,0 +1,120 @@
+fi
+
+# Create a wallet if it doesn't exist
+bitcoin-cli createwallet test 2>/dev/null
```
this might be useless. I believe that people running a bitcoind already have a wallet

solarsailorneo replied:
removing

0xEniotna commented on run_tests.sh:
```diff
@@ -0,0 +1,120 @@
+
+# Conditional block generation based on the network
+if [ "$NETWORK" == "regtest" ]; then
+    bitcoin-cli -regtest -generate 150
```
Idk if this is usefull. It is for sure when you launch a node but if the user already has a daemon then maybe not

solarsailorneo replied:
removing this line: `bitcoin-cli -regtest -generate 150`

0xEniotna commented on crates/bitcoin-da/tests/bitcoin_da_tests.rs:
```diff
@@ -357,40 +409,220 @@ mod tests {
+    
+        if cfg!(feature = "regtest") {
+            // Generate a block immediately in regtest
+            relayer.generate_blocks(1).unwrap();
```
Is there a rational to do it twice ?

solarsailorneo replied:
Not really. I made a mistake. Removed and tested. Have not pushed yet.

solarsailorneo replied:
pushed

0xEniotna commented on crates/bitcoin-da/src/lib.rs:
```diff
@@ -68,10 +71,19 @@ impl fmt::Display for BitcoinError {
             BitcoinError::ReadNoDataErr => write!(f, "Read no data in tx error"),
             BitcoinError::GetBlockErr => write!(f, "Get block error"),
             BitcoinError::GetBlockchainInfoErr => write!(f, "Get blockchain info error"),
+            BitcoinError::RpcError(ref e) => write!(f, "RPC error: {}", e),  // New match arm
```
This is SO USEFULL thanks, idk why I didn't implement it before

solarsailorneo replied:
No problem! Glad it helps

0xEniotna commented:
Wrote several new comments !
Otherwise it looks good, as soon as you address those, I'll merge

solarsailorneo committed:
Changed according to PR comments

+ Removed initializing of node from run_tests.sh.
+ Test_write test works for both regtest and signet with auto toggle from script or manual cargo command.
+ I had typo. Changed regnet to regtest in all areas.
+ Added back test_read_transaction and test_read_height.
+ Clarity provided in manual tests that fail in auto run.
+ Modified Cargo.toml files to remove warning during compiling.
+ Mod README. Added flag for long tests. Default does not run long tests.
+ Removed unnecessary things when running automatic tests (from PR comments).
+ Refactored test_read_data_by_block_height to include unique identifier as multiple writes can be written in the same block (from other tests).
+ Simplified functions. Polling mechanisms instead of sleep.
+ Clarified readme for long tests and automatic steps.

0xEniotna committed:
:construction: add small changes

0xEniotna commented:
@solarsailorneo One small last change ser, the user should be able to specify a bitcoin.conf path with `-conf=`

solarsailorneo committed:
pass --regtest or --signet to script

solarsailorneo committed:
More informative readme

0xEniotna commented:
Seems good to me !

0xEniotna committed:
remove stuff from script
